lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,313 @@
+            return self.reg.values
+
+    def _traverse_subkeys(self, stop_condition):
+        """Perform simple BFS of subkeys, returning the key
+        that successfully triggers the stop condition.
+        Args:
+            stop_condition: lambda or function pointer that takes a single argument
+                            a key and returns a boolean value based on that key
+        Return:
+            the key if stop_condition is triggered, or None if not
+        """
+        if not self._valid_reg_check():
+            return None
+        with self.invalid_reg_ref_error_handler():
+            queue = self.reg.subkeys
+            for key in queue:
+                if stop_condition(key):
+                    return key
+                queue.extend(key.subkeys)
+            return None
+
+    def find_subkey(self, subkey_name, recursive=True):
+        """If non recursive, this method is the same as get subkey with error handling
+        Otherwise perform a BFS of subkeys until desired key is found
+        Returns None or RegistryKey object corresponding to requested key name
+
+        Args:
+            subkey_name (str): string representing subkey to be searched for
+            recursive (bool): optional argument, if True, subkey need not be a direct
+                                sub key of this registry entry, and this method will
+                                search all subkeys recursively.
+                                Default is True
+        Return:
+            the desired subkey as a RegistryKey object, or none
+        """
+        if not recursive:
+            return self.get_subkey(subkey_name)
+
+        else:
+            return self._traverse_subkeys(lambda x: x.name == subkey_name)
+
+    def find_value(self, val_name, recursive=True):
+        """
+        If non recursive, return RegistryValue object corresponding to name
+
+        Args:
+            val_name (str): name of value desired from registry
+            recursive (bool): optional argument, if True, the registry is searched recursively
+                              for the value of name val_name, else only the current key is searched
+        Return:
+            The desired registry value as a RegistryValue object if it exists, otherwise, None
+        """
+        if not recursive:
+            return self.get_value(val_name)
+
+        else:
+            key = self._traverse_subkeys(lambda x: val_name in x.values)
+            if not key:
+                return None
+            else:
+                return key.values[val_name]
+
+
+def open_key(root, subkey):
+    """Returns registry key handle derived from root and its subkey
+    root can be either a WindowsRegistry or RegistryKey object
+
+    Note: If called on non Windows platforms, will raise RuntimeError
+    """
+    if not is_windows:
+        raise RuntimeError("Cannot invoke Windows registry methods on non Windows platform")
+    return root.get_subkey(subkey)
```

Comment:
IMO this and `get_value` below no longer add much: is the main thing that is raises an error? WindowsRegistryView already does that, although the RegistryKey initializer doesn't, so I could see that being useful here. If we move the `raise` into the initializer, I assume that would cover what these two functions intend.

Reply:
Agreed. They initially served a purpose but we refactored it away, removing.